### PR TITLE
Add Volo.Abp.BlobStoring.Aliyun

### PR DESCRIPTION
**Pull request by Aprite**
Aliyun OSS
Sub-account access to OSS 
https://help.aliyun.com/product/31815.html
STS temporary authorization to access OSS
https://help.aliyun.com/document_detail/100624.html

### CODE REVIEW

**Comment by liangshiw** [comment]
Hi @Aprite 

The unit tests error.

**Comment by Aprite** [comment]
@liangshiw  Because UserSecrets is not set
I have removed it

File: framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+            Configure<AbpBlobStoringOptions>(options =>
+            {
+                options.Containers.ConfigureAll((containerName, containerConfiguration) =>
+                {
+                    containerConfiguration.UseAliyun(aliyun =>
+                    {
+                        aliyun.AccessKeyId = _accessKeyId;
+                        aliyun.AccessKeySecret = _accessKeySecret;
+                        aliyun.Endpoint = _endpoint;//eg:https://oss-cn-beijing.aliyuncs.com
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs:46]: I don't think comment is necessary.
You should refer to the Azure Provider test module code,we should generate a random container and delete it after the test execution is end.

File: framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AliyunBlobContainer_Tests.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    //    private readonly IBlobContainer<MyTestContainer> _blobContainer;
+    //    private readonly string _blobname = "my-blob";
+    //    private readonly string _stringContent = "my-blob-content";
+    //    public AliyunBlobContainer_Tests()
+    //    {
+    //        _blobContainer = GetRequiredService<IBlobContainer<MyTestContainer>>();
+    //    }
+
+    //    [Fact]
+    //    public async void BlobContainer_Test()
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AliyunBlobContainer_Tests.cs:19]: I think this test code can be remove.

File: framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/MyTestContainer.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Volo.Abp.BlobStoring.Aliyun
+{
+    [BlobContainerName("my-container")]
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/MyTestContainer.cs:3]: Can remove.

File: framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                AssumeRoleRequest request = new AssumeRoleRequest
+                {
+                    AcceptFormat = FormatType.JSON,
+                    //eg:acs:ram::$accountID:role/$roleName
+                    RoleArn = aliyunConfig.RoleArn,
+                    RoleSessionName = aliyunConfig.RoleSessionName,
+                    //Set the validity period of the temporary access credential, the unit is s, the minimum is 900, and the maximum is 3600. default 3600
+                    DurationSeconds = aliyunConfig.DurationSeconds,
+                    //Set additional permission policy of Token; when acquiring Token, further reduce the permission of Token by setting an additional permission policy
+                    Policy = aliyunConfig.Policy.IsNullOrEmpty() ? null : aliyunConfig.Policy,
```
Comment (liangshiw) [inline framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs:65]: `Policy = aliyunConfig.Policy;`

File: framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                };
+                var response = client.GetAcsResponse(request);
+                iOssClient = new OssClient(aliyunConfig.Endpoint, response.Credentials.AccessKeyId, response.Credentials.AccessKeySecret, response.Credentials.SecurityToken);
+                _cache.Set(key, iOssClient, TimeSpan.FromSeconds(aliyunConfig.DurationSeconds));
```
Comment (liangshiw) [inline framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs:69]: Use `IDistributedCache` cache `AccessKeyId`, `AccessKeySecret` and `SecurityToken`

**Comment by liangshiw** [comment]
In addition, please delete the Chinese comment in the code. Thanks. : )

File: framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs (hunk 4 above)
Comment (Aprite) [reply framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs:65]: Policy = aliyunConfig.Policy; has bug with ""

File: framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                        aliyun.ContainerName = _randomContainerName;
+                        _configuration = aliyun;
+                    });
+                });
+            });
+        }
+
+        public override void OnApplicationShutdown(ApplicationShutdownContext context)
+        {
+            var ossClientFactory = context.ServiceProvider.GetService<IOssClientFactory>();
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs:67]:
```csharp
var objects = ossClient.ListObjects(_randomContainerName);

if (objects.ObjectSummaries.Any())
{
    ossClient.DeleteObjects(new DeleteObjectsRequest(_randomContainerName,
        objects.ObjectSummaries.Select(o => o.Key).ToList()));
}

ossClient.DeleteBucket(_randomContainerName);
```

File: framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+using Volo.Abp.DependencyInjection;
+
+namespace Volo.Abp.BlobStoring.Aliyun
+{
+    /// <summary>
+    /// Sub-account access to OSS or STS temporary authorization to access OSS
+    /// </summary>
+    public class DefaultOssClientFactory : IOssClientFactory, ITransientDependency
+    {
+        private readonly IDistributedCache<AssumeRoleCredentialsCacheItem> _cache;
```
Comment (liangshiw) [inline framework/src/Volo.Abp.BlobStoring.Aliyun/Volo/Abp/BlobStoring/Aliyun/DefaultOssClientFactory.cs:19]: Make `protected`

**Comment by liangshiw** [comment]
Unit test get exceptions.

image

File: framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        public override void ConfigureServices(ServiceConfigurationContext context)
+        {
+            context.Services.ReplaceConfiguration(ConfigurationHelper.BuildConfiguration(builderAction: builder =>
+            {
+                builder.AddUserSecrets(UserSecretsId);
+            }));
+
+            var configuration = context.Services.GetConfiguration();
+            var _accessKeyId = configuration["Aliyun:AccessKeyId"];
```
Comment (liangshiw) [inline framework/test/Volo.Abp.BlobStoring.Aliyun.Tests/Volo/Abp/BlobStoring/Aliyun/AbpBlobStoringAliyunTestModule.cs:35]: Remove `_`

**Comment by liangshiw** [comment]
Thanks a lot @Aprite  👍

**Pull request merged**